rust/uefi/linux-bootloader/src/companions.rs

```diff
@@ -0,0 +1,158 @@
+/// during discovery workflows, e.g. finding files in drop-in directories.
+pub enum CompanionInitrd {
+    Credentials(Cpio),
+    GlobalCredentials(Cpio),
+    SystemExtension(Cpio),
+    PcrSignature(Cpio),
+    PcrPublicKey(Cpio),
+}
+
+impl CompanionInitrd {
+    pub fn into_cpio(self) -> Cpio {
+        match self {
+            Self::Credentials(c)
+            | Self::GlobalCredentials(c)
+            | Self::SystemExtension(c)
+            | Self::PcrSignature(c)
+            | Self::PcrPublicKey(c) => c,
+        }
+    }
+}
+
+/// Discover any credentials, i.e. files ending by .cred
+/// Credentials comes into two variants:
+///     - global credentials ($ESP/loader/credentials/*.cred), global to the ESP
+///     - local credentials ($path_to_image.extra/*.cred), specific to this image
+///
+/// Those will be unmeasured, you are responsible for measuring them or not.
+/// But CPIOs are guaranteed to be stable and independent of file discovery order.
+pub fn discover_credentials(
+    fs: &mut uefi::fs::FileSystem,
+    default_dropin_dir: Option<&Path>,
+) -> uefi::Result<Vec<CompanionInitrd>> {
+    let mut companions = Vec::new();
+
+    let default_global_dropin_dir = cstr16!("\\loader\\credentials");
+    if fs.try_exists(default_global_dropin_dir).unwrap() {
+        let metadata = fs.metadata(default_global_dropin_dir).expect("Failed to obtain metadata on `\\loader\\credentials` path (which is supposed to exist)");
+        if metadata.is_directory() {
+            let global_credentials: Vec<PathBuf> =
+                find_files(fs, default_global_dropin_dir.as_ref(), ".cred")?;
+
+            if !global_credentials.is_empty() {
+                companions.push(CompanionInitrd::GlobalCredentials(
+                    pack_cpio(
+                        fs,
+                        global_credentials,
+                        ".extra/global_credentials",
+                        0o500,
+                        0o400,
+                    )
+                    .expect("Failed to pack CPIO"),
+                ));
+            }
+        }
+    }
+
+    if let Some(default_dropin_dir) = default_dropin_dir {
+        let local_credentials: Vec<PathBuf> = find_files(fs, default_dropin_dir, ".cred")?;
+
+        if !local_credentials.is_empty() {
+            companions.push(CompanionInitrd::Credentials(
+                pack_cpio(fs, local_credentials, ".extra/credentials", 0o500, 0o400)
+                    .expect("Failed to pack CPIO"),
+            ));
+        }
+    }
+
+    Ok(companions)
+}
+/// Discover any system image extension, i.e. files ending by .raw
+/// They must be present inside $path_to_image.extra/*.raw, specific to this image.
+///
+/// Those will be unmeasured, you are responsible for measuring them or not.
+/// But CPIOs are guaranteed to be stable and independent of file discovery order.
+pub fn discover_system_extensions(
+    fs: &mut uefi::fs::FileSystem,
+    default_dropin_dir: &Path,
```

Comment:
Why is this parameter optional above and not here?

Reply:
If there's no dropin directory for system extensions, there's no attempt to load them.
If there's no dropin directory for systemd credentials, we can still try to load them from a default dropin directory location.

We can make both of them take &Path and put the default inside of the caller site if we want but that's weird because the logic comes from the function itself.